### ISSUE TITLE
onwrite/generateBundle hook execution order not guaranteed any more

### ISSUE THREAD

**Issue by kryops** [rollup/rollup#2305]
Hi everyone,

it seems like the onwrite plugin hook does not execute in the order that the plugins were defined any more.

I got the following issue report today:
https://github.com/kryops/rollup-plugin-gzip/issues/2

This seems to be a regression from 0.60 - I followed it to this commit:
https://github.com/rollup/rollup/commit/a329517f309a3b6a2309a9240ed24cd6a251ed20

It breaks cases like the following:

- plugin 1 hooks into the `onwrite` callback, writes an additional file and then resolves the promise it returned
- plugin 2 hooks into the `onwrite` callback as well, and wants to do something with the file written by plugin 1

What is the intended behavior in this case?

- Should we be able to rely on the plugin definition order?
- Will all this be superseded by the new asset definition workflow anyway? (so plugin 1 will have to define its file as an asset)

Right now the only thing I could do in plugin 2 is to wait a certain amount of time and hope that plugin 1 is done :confused: 

Reproduction:
```
import * as fs from 'fs'

rollup
        .rollup({
            input: 'test/sample/index.js',
            plugins: [
                {
                    name: 'test1',
                    onwrite: (options, bundle) => {
                        return new Promise(resolve => {
                            setTimeout(() => {
                                fs.writeFile('test/__output/test.txt', 'This is a test', () => resolve());
                            }, 1000);
                        });
                    }
                },
                {
                    name: 'test2',
                    onwrite: (options, bundle) => {
                        return new Promise((resolve, reject) => {
                            fs.stat('test/__output/test.txt', (err, stats) => {
                                if(err) reject('File not written!');
                                else resolve();
                            });
                        });
                    }
                },
            ]
        })
        .then(bundle => {
            return bundle.write({
                file: 'test/__output/bundle.js',
                format: 'cjs'
            });
        })
```

**Comment by guybedford** [comment]
@kryops let me see if we can revert the behaviour so that it maintains compatibility here.

The new way recommended for this is through the transformBundle hook:

```js
generateBundle (outputOptions, bundle, isWrite) {
  Object.keys(bundle).forEach(id => {
    const chunk = bundle[id];
    { code: chunk.code, map: chunk.map } = minify(chunk.code, chunk.map);
  });
}
```

would that continue to solve your use case in future if onwrite is deprecated?

**Comment by kryops** [comment]
@guybedford thanks for the suggestion.

I tried adding the `generateBundle` hook to support code splitting as well:
https://github.com/kryops/rollup-plugin-gzip/commit/7d21c5cd096fc1ce1c1b9f08c44d8d4fa5437726

In contrast to the `onwrite` hook it seems to be called before the files are actually written though, even with the `isWrite` flag being set.

In my plugin, I want to use the files written on disk to create gz-compressed versions of them in addition to the original bundle/chunk files. I tried using `bundle.code` (or now `chunk.code`) before to achieve this, but that wasn't identical to what actually ended up being written on disk because it was missing the sourceMap comment.

So what I need is ideally a hook which is executed after all chunks are written on disk, just like what the old `onwrite` hook does ;)

**Comment by guybedford** [comment]
@kryops one of the design goals of the plugin hooks was to ensure that plugins could work in-memory allowing build workflows that don't touch the filesystem.

The `generateBundle` provides an exact image of what will be written to disk before it is written to disk, and anything added to that object can set new files / modify the existing ones.

Does the above still cater to your use case?

**Comment by kryops** [comment]
that actually sounds pretty good - I'm away on holidays now, but I will try it out after I come back!

**Comment by kryops** [comment]
Even with the `generateBundle` hook I still have the problem that `chunk.code` does not include the source map comment (see below). Though I can probably work around it by adding it manually if `chunk.map` is set. Will continue to experiment...

Chunk object:
```
{ 'bundle.js':
   { imports: [],
     exports: [],
     modules:
      { '/home/michael/git/rollup-plugin-gzip/test/sample/foobar.js': [Object],
        '/home/michael/git/rollup-plugin-gzip/test/sample/index.js': [Object] },
     code:
      '(function () {\n    \'use strict\';\n\n    function foobar() {\n        console.log(\'foobar\');\n    }\n\n    foobar();\n\n}());\n',
     map:
      SourceMap {
        version: 3,
        file: 'bundle.js',
        sources: [Array],
        sourcesContent: [Array],
        names: [],
        mappings:
         ';;;IAAO,SAAS,MAAM,GAAG;IACzB,IAAI,OAAO,CAAC,GAAG,CAAC,QAAQ,EAAC;IACzB,CAAC;;ICAD,MAAM,EAAE;;;;' } } }
```
File written:
```
(function () {
    'use strict';

    function foobar() {
        console.log('foobar');
    }

    foobar();

}());
//# sourceMappingURL=bundle.js.map
```

**Comment by guybedford** [comment]
@kryops how did those experiments work out there?

**Comment by kryops** [comment]
Sorry, I haven't had too much time recently to find a perfect working solution.
I could probably work around the source map issue for my use case by partially reimplementing `writeOutputFile()`, but then I got stuck trying to add the compressed contents as new files to the bundle. I tried using the `emitAsset()` function, but that would write the file to `assets/bundle.js-460a2d0d.gz` instead of `bundle.js.gz`.

This is what I ended up with before I gave up:

```ts
import { Plugin } from 'rollup'
import { gzip, ZlibOptions } from 'zlib'

export interface GzipPluginOptions {
    algorithm?: string
    options?: ZlibOptions
    additional?: string[]
    minSize?: number
    delay?: number
}

function gzipPlugin(_options: GzipPluginOptions = {}): Plugin {
    const plugin: Plugin = {
        name: 'gzip',

        generateBundle(_outputOptions, bundle, isWrite) {
            if (!isWrite) return

            return Promise.all(
                Object.keys(bundle).map(chunkId => {
                    const chunk = bundle[chunkId] as any

                    // TODO source map comment
                    const chunkContent: Buffer = Buffer.isBuffer(chunk)
                        ? (chunk as Buffer)
                        : chunk.code !== undefined
                            ? chunk.code
                            : Buffer.from(bundle[chunkId] as string)

                    return new Promise((resolve, reject) => {
                        gzip(chunkContent, (err, result) => {
                            if (err) {
                                reject(
                                    '[rollup-plugin-gzip] Error compressing file ' +
                                        chunkId,
                                )
                            } else {
                                // TODO writes the file to the wrong location
                                this.emitAsset(chunkId + '.gz', result)
                                resolve()
                            }
                        })
                    })
                }),
            ) as Promise<any>
        },
    }

    return plugin
}

export default gzipPlugin
```

**Comment by kryops** [comment]
Aah, that sounds simple enough then, thanks!
I'm on a business trip at the moment, will continue to experiment when I find time again.

**Comment by kryops** [comment]
Hi @guybedford,
thanks for all your support! I just released v2.0.0 of `rollup-plugin-gzip` using the `generateBundle` hook.

So we can finally return this issue to its original topic, which is the execution order of the `onwrite` and `generateBundle` hooks :smile:

**Comment by guybedford** [comment]
@kryops glad to hear you made some progress there! Could you explain to me once again the use case for still needing `onwrite` here? The plan was to deprecate this hook. You mentioned you need `onwrite` to access the source map comment? Or was it something else?

**Comment by kryops** [comment]
Hi @guybedford,
in my new implementation I don't use the `onwrite` hook any more. Instead I partially copied the `writeOutputFile()` function from rollup - the part that creates the final source string from the code and the source map: https://github.com/kryops/rollup-plugin-gzip/blob/master/src/index.ts#L81

**Comment by kryops** [comment]
yes, probably - and as long as the `onWrite` hook is still around it would probably also be good to have a defined execution order there as well for backwards compatibility :wink:

**Comment by guybedford** [comment]
Yes, good point, it's definitely a regression we can mitigate - I've created a fix in https://github.com/rollup/rollup/pull/2364.

**Comment by guybedford** [comment]
This was merged in #2364.
